### DISCUSSION TITLE
Which input variables should we add to improve from baseline?

### DISCUSSION THREAD

**Discussion by jsadler2** [USGS-R/drb-do-ml#44] [Questions]
In our baseline model (#40) we use the following x variables:

![image](https://user-images.githubusercontent.com/6943441/151255846-ea93a0e8-f042-4ac6-8961-cf58895ec05e.png)

as specified in the config file:
 https://github.com/USGS-R/drb-do-ml/blob/9a45f24fe7fed2ff8c9ad9404052e71e61524302/2a_model/src/models/0_baseline_LSTM/config.yml#L5

## But...
These are probably not _the_ ideal set of input variables. 

## So
What should we add?

**Comment by jsadler2** [comment]
## Discharge
One that @amcarter brought up was discharge.

One challenge with this that only 8 of the 13 sites we are using have any discharge data at all:
```
"01466500" "01473500" "01474500" "01475530" "01475548" "01480617" "01480870" "01481000"
```
and only 5 of those 8 have we been using for training:
![image](https://user-images.githubusercontent.com/6943441/151257589-f528df10-1fb7-4e48-9fcc-01c6bfb6e90d.png)

**Reply by jsadler2** [reply]
One possibility is to us PGDL predictions of discharge as input.

**Reply by jsadler2** [reply]
Another thing that has been brought up is that maybe the model is learning "discharge" from the input data (e.g., precip and catchment attributes) already. (see #45 )

**Reply by amcarter** [reply]
I think that keeping discharge out of the model in the hopes of creating a more generally applicable model is a good plan. It seems likely that the model is learning something that looks a lot like discharge from the inputs that are already included. But, because of the importance of discharge in controlling DO, it seems worth it to include as many of the input parameters used in the[ Kratzert et al LSTM models of discharge ](https://agupubs.onlinelibrary.wiley.com/doi/pdfdirect/10.1029/2019WR026065)that we can easily get. Some possibilities that seem easy to start with below.

Dynamic variables:
- daily min and max air temp instead of average (this would likely be good for predicting DO min and max in addition to discharge)
- vapor pressure
- solar radiation (duplicates cloud cover to an extent, but might add new information)

Static variables:
- Mean daily precip (as a static variable in addition to the dynamic one)
- Mean daily PET (potential evapotranspiration)
These might be harder to get, but seem really useful:
- soil depth
- soil porosity
- soil conductivity
- geological permeability

**Reply by amcarter** [reply]
Also, Kratzert uses lots of calculated metrics from daily precip that wouldn't be difficult to generate, but I don't have a good understanding of what that adds to the model's ability to learn, given that it had that data already. I also don't know how much more difficult it makes the model to run as you add more parameters. I'd be interested to hear what you all think about that.

**Reply by galengorski** [reply]
It seems like because the model has trouble generalizing in space, adding in static attributes that vary as much as possible across the sites might produce the best results. For these variables that we think may have an influence (but we aren't certain) it might make sense to collect them and look at the standard deviation (or entropy) across the sites, particularly looking at differences between training and testing sites.

**Reply by jsadler2** [reply]
@amcarter - I agree that adding all of those variables would be useful.

We will soon have all of the gridmet dynamic drivers available to us. These include min/max temp and solar radiation. It does not include vapor pressure, but it does include min/max relative humidity which is related.

@lekoenig is working on gathering static variables for another project and there are _many_ available including PET and precip. Some of these have to do with the soils [see here](https://www.sciencebase.gov/catalog/file/get/5728decfe4b0b13d3918a9aa?f=__disk__c4%2F36%2F6f%2Fc4366f16b25a57a8ff9cf7da9ae29159f991b058&transform=1&allowOpen=true). Of these, there are several soil parameters available. Here are a few that may get us toward the characteristics you listed:
Name|DESCRIPTION|UNITS
---|---|---
PERMAVE|Average value for the range in permeability (inches per hour) per NHDPlus version 2 catchment.|inches per hour
RFACT|Average Rainfall Runoff Factor from Universal Soil Loss Equation per NHDPlus version 2 catchment.|100s ft-tonf in/h/ac/yr
ROCKDEP|Average value for the range in the total soil thickness examined (inches) per NHDPlus version 2 catchment.|inches
BDAVE|Average value for the value of bulk density (grams per cubic centimeter) per NHDPlus version 2 catchment.|grams per unit centimeter
WTDEP|Average value for the range in depth to the seasonally high water table (feet) per NHDPlus version 2 catchment.|feet
AWCAVE|Average value for the range in available water capacity (fraction) per NHDPlus version 2 catchment.|inches of water per inches of soil depth
CLAYAVE|Average percent of clay in soil per NHDPlus version 2 catchment. -9999 denotes NODATA usually water.|NA
SANDAVE|Average percent of sand in soil per NHDPlus version 2 catchment. -9999 denotes NODATA usually water.|NA
SILTAVE|Average percent of silt in soil per NHDPlus version 2 catchment. -9999 denotes NODATA usually water.|NA

**Reply by jsadler2** [reply]
> Also, Kratzert uses lots of calculated metrics from daily precip that wouldn't be difficult to generate, but I don't have a good understanding of what that adds to the model's ability to learn, given that it had that data already. I also don't know how much more difficult it makes the model to run as you add more parameters. I'd be interested to hear what you all think about that.

Because LSTM's have the `c` and `h` internal states that let things "accumulate," and then "release," I'm inclined to hope that the model is figuring out on its own those "calculated metrics" that we _could_ provide. That said, it would be nearly trivial to provide those as inputs if that's what we decide.

**Reply by amcarter** [reply]
Those soil characteristics look great, especially the first two. If all of them are readily available they would all probably be informative for discharge. I would say include all of them unless there is a big computational cost associated with that.

**Comment by jsadler2** [comment]
@galengorski regarding your comment (i'm starting this in a new "thread" so that the other can be more geared to discharge):

> It seems like because the model has trouble generalizing in space, adding in static attributes that vary as much as possible across the sites might produce the best results. For these variables that we think may have an influence (but we aren't certain) it might make sense to collect them and look at the standard deviation (or entropy) across the sites, particularly looking at differences between training and testing sites.

I agree that this is a good idea. I have done a first look at this with the 8 static features that we have:
![image](https://user-images.githubusercontent.com/6943441/151844128-222ccc4d-0c68-4e76-ad87-9f1e250cf777.png)

**Comment by lekoenig** [comment]
Mean air temperature is currently included as an input feature (`seg_tave_air`) but water temperature is not. PRMS outputs include predicted water temperature (`seg_tave_water`). Perhaps we should add that variable to our list of dynamic input features.
